crates/rccheck/src/lib.rs

```diff
@@ -57,6 +62,118 @@ pub trait Certifiable {
     fn public_key_to_spki(public_key: &Self::PublicKey) -> Vec<u8>;
 }
 
+/// A collection of X.509 `SubjectPublicKeyInfo` (SPKI)
+///
+/// Implements the traits ClientCertVerifier and ServerCertVerifier, enabling the verification
+/// of X.509 certificates signed by a keypair in the set of Pre-Shared Keys (PSKs).
+///
+/// Example
+/// ```
+/// use rccheck::*;
+/// let mut rng = rand::thread_rng();
+/// let keypairs: Vec<_> = (0..10).into_iter().map(|_| ed25519_dalek::Keypair::generate(&mut rng)).collect();
+/// let spkis = keypairs.iter().map(|kp| ed25519_certgen::Ed25519::public_key_to_spki(&kp.public)).collect::<Vec<_>>();
+/// let pskset = PskSet::from_der(&spkis.iter().map(|spki| &spki[..]).collect::<Vec<_>>()[..]).unwrap();
+/// ```
+///
+#[derive(Debug, PartialEq, Serialize, Deserialize)]
+pub struct PskSet {
+    pub spki_set: BTreeSet<Psk>,
+}
+
+impl PskSet {
+    /// This creates a new `PskSet` from a set of X.509 `SubjectPublicKeyInfo` (SPKI) bytes (in DER format)
+    pub fn from_der(bytes_array: &[&[u8]]) -> Result<PskSet, anyhow::Error> {
+        let set = bytes_array
+            .iter()
+            .map(|&bytes| {
+                let psk = Psk::from_der(bytes)?;
+                Ok(psk)
+            })
+            .collect::<Result<BTreeSet<Psk>, anyhow::Error>>()?;
+        Ok(PskSet { spki_set: set })
+    }
+}
+
+/// A `ClientCertVerifier` that will ensure that every client provides a valid, expected
+/// certificate, without any name checking.
+impl ClientCertVerifier for PskSet {
+    fn offer_client_auth(&self) -> bool {
+        true
+    }
+
+    fn client_auth_mandatory(&self) -> Option<bool> {
+        Some(true)
+    }
+
+    fn client_auth_root_subjects(&self) -> Option<rustls::DistinguishedNames> {
+        // We can't guarantee subjects before having seen the cert. This should not be None for compatiblity reasons
+        Some(rustls::DistinguishedNames::new())
+    }
+
+    // Verifies this is a valid certificate self-signed by a public key we expect(in PSK)
+    // 1. We check if the key exists in the set of expected keys.
+    // 2. Verify the key against expected key.
+    fn verify_client_cert(
+        &self,
+        end_entity: &rustls::Certificate,
+        intermediates: &[rustls::Certificate],
+        now: SystemTime,
+    ) -> Result<ClientCertVerified, rustls::Error> {
+        // 1. Check if the key exists in the set of expected keys.
+        let cert = X509Certificate::from_der(&end_entity.0[..])
+            .map_err(|_| rustls::Error::InvalidCertificateEncoding)?;
+        let pk_bytes = cert.1.public_key().raw.to_vec();
+        let spki = Psk::from_der(&pk_bytes).unwrap();
```

Comment:
Wondering if it would make sense to have a set of the DER serialized key formats Vs a set of  actual PKs. That would allow for not requiring to extract the key from its DER format saving one deserialization operation.

Reply:
It does save a lot of deserialization. I wonder why we didn't do this for the original verifier too though - @huitseeker?